server/src/test/java/org/elasticsearch/common/file/MasterNodeFileWatchingServiceTests.java

```diff
@@ -0,0 +1,88 @@
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.gateway.GatewayService;
+import org.elasticsearch.node.NodeRoleSettings;
+import org.elasticsearch.test.ESTestCase;
+import org.junit.After;
+import org.junit.Before;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.concurrent.ExecutionException;
+
+import static org.hamcrest.Matchers.is;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class MasterNodeFileWatchingServiceTests extends ESTestCase {
+
+    MasterNodeFileWatchingService testService;
+    Path watchedFile;
+    Runnable fileChangedCallback;
+
+    @Before
+    public void setupTestService() throws IOException {
+        watchedFile = createTempFile();
+        ClusterService clusterService = mock(ClusterService.class);
+        Settings settings = Settings.builder()
+            .put(NodeRoleSettings.NODE_ROLES_SETTING.getKey(), DiscoveryNodeRole.MASTER_ROLE.roleName())
+            .build();
+        when(clusterService.getSettings()).thenReturn(settings);
+        fileChangedCallback = () -> {};
+        testService = new MasterNodeFileWatchingService(clusterService, watchedFile) {
+
+            @Override
+            protected void processFileChanges() throws InterruptedException, ExecutionException, IOException {
+                fileChangedCallback.run();
+            }
+
+            @Override
+            protected void processInitialFileMissing() throws InterruptedException, ExecutionException, IOException {
+                // file always exists, but we don't care about the missing case for master node behavior
+            }
+        };
+        testService.start();
+    }
+
+    @After
+    public void stopTestService() {
+        testService.stop();
+    }
+
+    public void testBecomingMasterNodeStartsWatcher() {
+        final DiscoveryNode localNode = DiscoveryNodeUtils.create("node");
+        ClusterState notRecoveredClusterState = ClusterState.builder(ClusterName.DEFAULT)
+            .nodes(DiscoveryNodes.builder().add(localNode).localNodeId(localNode.getId()).masterNodeId(localNode.getId()))
+            .blocks(ClusterBlocks.builder().addGlobalBlock(GatewayService.STATE_NOT_RECOVERED_BLOCK))
+            .build();
+        testService.clusterChanged(new ClusterChangedEvent("test", notRecoveredClusterState, ClusterState.EMPTY_STATE));
+        // just a master node isn't sufficient, cluster state also must be recovered
+        assertThat(testService.watching(), is(false));
+
+        ClusterState recoveredClusterState = ClusterState.builder(notRecoveredClusterState)
+            .blocks(ClusterBlocks.EMPTY_CLUSTER_BLOCK)
+            .build();
+        testService.clusterChanged(new ClusterChangedEvent("test", recoveredClusterState, notRecoveredClusterState));
+        // just a master node isn't sufficient, cluster state also must be recovered
+        assertThat(testService.watching(), is(true));
```

Comment:
Maybe also verify that we stop watching if either of those conditions becomes `false` again?